File: examples/python/task-management/api-service/src/external/notification_queue.py
"""Redis queue for background notifications."""
import redis
import json
import os
from typing import Dict, Any

class NotificationQueue:
    """Queue for background notification jobs."""

    def __init__(self):
        redis_url = os.getenv("REDIS_URL", "redis://localhost:6379/0")
        self.client = redis.from_url(redis_url)
        self.queue_name = "notifications"

    def send_task_created(self, task_id: int, user_id: int) -> None:
        """Queue notification for task creation."""
        message = {
            "type": "task_created",
            "task_id": task_id,
            "user_id": user_id
        }
        self.client.rpush(self.queue_name, json.dumps(message))

    def send_task_reminder(self, task_id: int, user_id: int) -> None:
        """Queue notification for task reminder."""
        message = {
            "type": "task_reminder",
            "task_id": task_id,
            "user_id": user_id
        }
        self.client.rpush(self.queue_name, json.dumps(message))

File: examples/csharp/task-management/api-service/src/External/NotificationQueue.cs
using System;
using System.Net.Http;
using System.Text;
using System.Text.Json;
using System.Threading.Tasks;

namespace TaskManagementApi.External;

/// <summary>
/// Client for sending messages to notification queue.
/// </summary>
public class NotificationQueue
{
    private readonly HttpClient _httpClient;
    private readonly ILogger<NotificationQueue> _logger;
    private readonly string _queueUrl;

    public NotificationQueue(HttpClient httpClient, ILogger<NotificationQueue> logger, IConfiguration configuration)
    {
        _httpClient = httpClient;
        _logger = logger;
        _queueUrl = configuration["NotificationQueue:Url"] ?? string.Empty;
    }

    public async Task<bool> EnqueueTaskCreatedAsync(int taskId, int userId, string taskTitle)
    {
        _logger.LogInformation("Enqueueing task created notification for task {TaskId}", taskId);

        var message = new
        {
            eventType = "task.created",
            taskId,
            userId,
            taskTitle,
            timestamp = DateTime.UtcNow
        };

        return await SendMessageAsync(message);
    }

    public async Task<bool> EnqueueTaskCompletedAsync(int taskId, int userId)
    {
        _logger.LogInformation("Enqueueing task completed notification for task {TaskId}", taskId);

        var message = new
        {
            eventType = "task.completed",
            taskId,
            userId,
            timestamp = DateTime.UtcNow
        };

        return await SendMessageAsync(message);
    }

    private async Task<bool> SendMessageAsync(object message)
    {
        var content = new StringContent(
            JsonSerializer.Serialize(message),
            Encoding.UTF8,
            "application/json");

        try
        {
            var response = await _httpClient.PostAsync($"{_queueUrl}/messages", content);
            return response.IsSuccessStatusCode;
        }
        catch (HttpRequestException ex)
        {
            _logger.LogError(ex, "Failed to send message to notification queue");
            return false;
        }
    }
}
